ssl/ech/ech_internal.c

```diff
@@ -160,4 +214,1260 @@ int ossl_ech_conn_init(SSL_CONNECTION *s, SSL_CTX *ctx,
+        WPACKET_cleanup(&tpkt);
+        BUF_MEM_free(tpkt_mem);
+    }
+    EVP_MD_CTX_free(ctx);
+    return 0;
+}
+
+/*
+ * reset the handshake buffer for transcript after ECH is good
+ * buf is the data to put into the transcript (inner CH if no HRR)
+ * blen is the length of buf
+ * return 1 for success
+ */
+int ossl_ech_reset_hs_buffer(SSL_CONNECTION *s, const unsigned char *buf,
+                             size_t blen)
+{
+# ifdef OSSL_ECH_SUPERVERBOSE
+    ossl_ech_pbuf("RESET transcript to", buf, blen);
+# endif
+    if (s->s3.handshake_buffer != NULL) {
+        (void)BIO_set_close(s->s3.handshake_buffer, BIO_CLOSE);
+        BIO_free(s->s3.handshake_buffer);
+        s->s3.handshake_buffer = NULL;
+    }
+    EVP_MD_CTX_free(s->s3.handshake_dgst);
+    s->s3.handshake_dgst = NULL;
+    s->s3.handshake_buffer = BIO_new(BIO_s_mem());
+    if (s->s3.handshake_buffer == NULL)
+        return 0;
+    /* providing nothing at all is a real use (mid-HRR) */
+    if (buf != NULL && blen > 0)
+        BIO_write(s->s3.handshake_buffer, (void *)buf, (int)blen);
+    return 1;
+}
+
+/*
+ * To control the number of zeros added after an EncodedClientHello - we pad
+ * to a target number of octets or, if there are naturally more, to a number
+ * divisible by the defined increment (we also do the spec-recommended SNI
+ * padding thing first)
+ */
+# define OSSL_ECH_PADDING_TARGET 128 /* ECH cleartext padded to at least this */
+# define OSSL_ECH_PADDING_INCREMENT 32 /* ECH padded to a multiple of this */
+
+/*
+ * figure out how much padding for cleartext (on client)
+ * ee is the chosen ECHConfig
+ * return overall length to use including padding or zero on error
+ *
+ * "Recommended" inner SNI padding scheme as per spec (section 6.1.3)
+ * Might remove the mnl stuff later - overall message padding seems
+ * better really, BUT... we might want to keep this if others (e.g.
+ * browsers) do it so as to not stand out compared to them.
+ *
+ * The "+ 9" constant below is from the specifiation and is the
+ * expansion comparing a string length to an encoded SNI extension.
+ * Same is true of the 31/32 formula below.
+ *
+ * Note that the AEAD tag will be added later, so if we e.g. have
+ * a padded cleartext of 128 octets, the ciphertext will be 144
+ * octets.
+ */
+static size_t ech_calc_padding(SSL_CONNECTION *s, OSSL_ECHSTORE_ENTRY *ee)
+{
+    int length_of_padding = 0, length_with_snipadding = 0;
+    int innersnipadding = 0, length_with_padding = 0;
+    size_t mnl = 0, clear_len = 0, isnilen = 0;
+
+    if (s == NULL || ee == NULL)
+        return 0;
+    mnl = ee->max_name_length;
+    if (mnl != 0) {
+        /* do weirder padding if SNI present in inner */
+        if (s->ext.hostname != NULL) {
+            isnilen = strlen(s->ext.hostname) + 9;
+            innersnipadding = mnl - isnilen;
+        } else {
+            innersnipadding = mnl + 9;
+        }
+        if (innersnipadding < 0)
+            innersnipadding = 0;
```

Comment:
This can be removed with my suggestion above

Reply:
yep